File: gamecore-core/src/main/java/net/malfact/gamecore/Vector3.java
package net.malfact.gamecore;

import org.jetbrains.annotations.NotNull;

import java.util.Random;
import java.util.function.DoubleBinaryOperator;
import java.util.function.DoubleUnaryOperator;

@SuppressWarnings("unused")
public class Vector3 {

    public static final Vector3 Zero = new Vector3(0, 0, 0);
    public static final Vector3 Up = new Vector3(0, 1.0, 0);
    public static final Vector3 Down = new Vector3(0, -1.0, 0);

    public final double x;
    public final double y;
    public final double z;

    public Vector3(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public int blockX() {
        return (int) Math.floor(x);
    }

    public int blockY() {
        return (int) Math.floor(y);
    }

    public int blockZ() {
        return (int) Math.floor(z);
    }

    @NotNull
    public Vector3 copy() {
        return new Vector3(x, y, z);
    }

    @NotNull
    public Vector3 add(@NotNull Vector3 other) {
        return new Vector3(
            this.x + other.x,
            this.y + other.y,
            this.z + other.z
        );
    }

    @NotNull
    public Vector3 subtract(@NotNull Vector3 other) {
        return new Vector3(
            this.x - other.x,
            this.y - other.y,
            this.z - other.z
        );
    }

    @NotNull
    public Vector3 multiply(@NotNull Vector3 other) {
        return new Vector3(
            this.x * other.x,
            this.y * other.y,
            this.z * other.z
        );
    }

    @NotNull
    public Vector3 multiply(double scalar) {
        return new Vector3(
            this.x * scalar,
            this.y * scalar,
            this.z * scalar
        );
    }

    @NotNull
    public Vector3 divide(@NotNull Vector3 other) {
        return new Vector3(
            this.x / other.x,
            this.y / other.y,
            this.z / other.z
        );
    }

    @NotNull
    public Vector3 divide(double scalar) {
        return new Vector3(
            this.x / scalar,
            this.y / scalar,
            this.z / scalar
        );
    }

    @NotNull
    public Vector3 negate() {
        return this.multiply(-1.0);
    }

    public double lengthSquared() {
        return (x * x) + (y * y) + (z * z);
    }

    public double length() {
        return Math.sqrt(lengthSquared());
    }

    public double distanceSquared(@NotNull Vector3 other) {
        return Math.pow(this.x - other.x, 2) + Math.pow(this.y - other.y, 2) + Math.pow(this.z - other.z, 2);
    }

    public double distance(@NotNull Vector3 other) {
        return Math.sqrt(distanceSquared(other));
    }

    public double dot(@NotNull Vector3 other) {
        return this.x * other.x + this.y * other.y + this.z + other.z;
    }

    public float angle(@NotNull Vector3 other) {
        return (float) Math.acos(dot(other));
    }

    @NotNull
    public Vector3 midpoint(@NotNull Vector3 other) {
        return new Vector3(
            (x + other.x) / 2.0,
            (y + other.y) / 2.0,
            (z + other.z) / 2.0
        );
    }

    @NotNull
    public Vector3 crossProduct(@NotNull Vector3 other) {
        return new Vector3(
            y * other.z - other.y * z,
            z * other.x - other.z * x,
            x * other.y - other.x * y
        );
    }

    @NotNull
    public Vector3 floor() {
        return unaryOp(this, Math::floor);
    }

    @NotNull
    public Vector3 ceil() {
        return unaryOp(this, Math::ceil);
    }

    @NotNull
    public Vector3 normalize() {
        return this.divide(length());
    }

    @Override
    public String toString() {
        return "<" + x + "," + y + "," + z + ">";
    }

    @NotNull
    public static Vector3 max(@NotNull Vector3 vec1, @NotNull Vector3 vec2) {
        return binaryOp(vec1, vec2, Math::max);
    }

    @NotNull
    public static Vector3 min(@NotNull Vector3 vec1, @NotNull Vector3 vec2) {
        return binaryOp(vec1, vec2, Math::min);
    }

    private static final Random random = new Random();

    @NotNull
    public static Vector3 random() {
        return new Vector3(random.nextDouble(), random.nextDouble(), random.nextDouble());
    }

    @NotNull
    public static Vector3 unaryOp(@NotNull Vector3 vec, DoubleUnaryOperator operator) {
        return new Vector3(
            operator.applyAsDouble(vec.x),
            operator.applyAsDouble(vec.y),
            operator.applyAsDouble(vec.z)
        );
    }

    @NotNull
    public static Vector3 binaryOp(@NotNull Vector3 vec1, @NotNull Vector3 vec2, DoubleBinaryOperator operator) {
        return new Vector3(
            operator.applyAsDouble(vec1.x, vec2.x),
            operator.applyAsDouble(vec1.y, vec2.y),
            operator.applyAsDouble(vec1.z, vec2.z)
        );
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Vector3 vec))
            return false;

        double e = 0.000001;
        return Math.abs(x - vec.x) < e && Math.abs(y - vec.y) < e && Math.abs(z - vec.z) < e;
    }
}

File: src/docs/lua/library/Vector3.lua
---@meta

--[[ Library ]] --

---The Vector3 library.
Vector3 = {}

---@type Vector3
Vector3.Zero = Vector3.new(0,0,0)
---@type Vector3
Vector3.Up = Vector3.new(0,1,0)
---@type Vector3
Vector3.Down = Vector3.new(0,-1,0)

---@param x number
---@param y number
---@param z number
---@return Vector3
function Vector3.new(x,y,z) end

---@return Vector3
function Vector3.random() end

--[[ Class ]] --

---@class Vector3: userdata
---@field x number
---@field y number
---@field z number
---@field blockX integer
---@field blockY integer
---@field blockZ integer
---@operator add(Vector3): Vector3
---@operator sub(Vector3): Vector3
---@operator mul(Vector3 | number): Vector3
---@operator div(Vector3 | number): Vector3
---@operator unm: Vector3
---@operator len: number
local instance = {}

---Clone the Vector3.
---@return Vector3
function instance.clone(self) end

---Get the length of the Vector3.
---@return number
function instance.length(self) end

---Get the square length of the Vector3.
---@return number
function instance.lengthSquared(self) end

---Get the Vector3 multiplied by `-1`.
---@return Vector3
function instance.neg(self) end

---Normalize the Vector3.
---@return Vector3
function instance.normalize(self) end

---Convert the Vector3 to a string.
---@return string
function instance.toString(self) end

---Get the Vector3 with the coordinates floored.
---@return Vector3
function instance.floor(self) end

---Get the Vector3 with the coordinates ceiled.
---@return Vector3
function instance.ceil(self) end

---Get the distance between this and another Vector3.
---@param other Vector3
---@return number
function instance.distance(self, other) end

---Get the square distance between this and another Vector3.
---@param other Vector3
---@return number
function instance.distanceSquared(self, other) end

---Get the midpoint between this and another Vector3.
---@param other Vector3
---@return Vector3
function instance.mid(self, other) end

---Get the cross product of this and another Vector3.
---@param other Vector3
---@return Vector3
function instance.cross(self, other) end

---Get the angle between this and another Vector3.
---@param other Vector3
---@return number
function instance.angle(self, other) end

---Get the dot product of this and another Vector3.
---@param other Vector3
---@return number
function instance.dot(self, other) end

---Add this and another Vector3.
---@param other Vector3
---@return Vector3
function instance.add(self, other) end

---Subtract this and another Vector3.
---@param other Vector3
---@return Vector3
function instance.sub(self, other) end

---Get the *(element-wise)* miximum between this and another Vector3.
---@param other Vector3
---@return Vector3
function instance.max(self, other) end

---Get the *(element-wise)* minimum between this and another Vector3.
---@param other Vector3
---@return Vector3
function instance.min(self, other) end

---Multiply this *(element-wise)* and another Vector3. 
---<br> **or** <br>
---Multiply this by a number.
---@param other Vector3 | number
---@return Vector3
function instance.multiply(self, other) end

---Divide this *(element-wise)* and another Vector3. 
---<br> **or** <br>
---Divide this by a number.
---@param other Vector3 | number
---@return Vector3
function instance.divide(self, other) end

---Check if this is equal to another value.
---@param other any
---@return boolean
function instance.equals(self, other) end

Vector3.clone = instance.clone
Vector3.length = instance.length
Vector3.lengthSquared = instance.lengthSquared
Vector3.neg = instance.neg
Vector3.normalize = instance.normalize
Vector3.toString = instance.toString
Vector3.floor = instance.floor
Vector3.ceil = instance.ceil
Vector3.distance = instance.distance
Vector3.distanceSquared = instance.distanceSquared
Vector3.mid = instance.mid
Vector3.cross = instance.cross
Vector3.angle = instance.angle
Vector3.dot = instance.dot
Vector3.add = instance.add
Vector3.sub = instance.sub
Vector3.max = instance.max
Vector3.min = instance.min
Vector3.multiply = instance.multiply
Vector3.divide = instance.divide
Vector3.equals = instance.equals